### PR TITLE
[halide-backend] Dimension-based indexing

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #129321
* #129320
* #129036
* #127506
* __->__ #129026
* #129025
* #126417

Prior to this the generated Halide code was a rather literal translation of the Triton code, with XBLOCK/YBLOCK/RBLOCK and 1D inputs.  Halide prefers dimensions, and this 1D index triggers a lot of bugs and perf issues.  This PR infers dimensions and changes the indexing in the generated code.

Before
```py
@hl.generator(name="kernel")
class Kernel:
    in_ptr0 = hl.InputBuffer(hl.Float(32), 1)
    out_ptr3 = hl.OutputBuffer(hl.Float(32), 2)

    def generate(g):
        in_ptr0 = g.in_ptr0
        out_ptr3 = g.out_ptr3
        xindex = hl.Var('xindex')
        rindex = hl.Var('rindex')
        r1 = rindex
        x0 = xindex
        idom = hl.RDom([hl.Range(0, 16), hl.Range(0, 32)])
        odom = hl.RDom([hl.Range(0, 16)])
        rdom = hl.RDom([hl.Range(0, 32)])
        xindex_idom = idom.x
        xindex_odom = odom.x
        rindex_idom = idom.y
        r1_idom = rindex_idom
        x0_idom = xindex_idom
        x0_odom = xindex_odom
        tmp0 = hl.Func('tmp0')
        tmp0[rindex, xindex] = in_ptr0[r1 + (32*x0)]
        tmp1 = hl.Func('tmp1')
        tmp1[xindex] = hl.maximum(rdom, tmp0[rdom, xindex])
        tmp2 = hl.Func('tmp2')
        tmp2[rindex, xindex] = tmp0[rindex, xindex] - tmp1[xindex]
        tmp3 = hl.Func('tmp3')
        tmp3[rindex, xindex] = hl.fast_exp(hl.cast(hl.Float(32), tmp2[rindex, xindex])) if tmp2.type().bits() <= 32 else hl.exp(tmp2[rindex, xindex])
        tmp4 = hl.Func('tmp4')
        tmp4[xindex] = hl.sum(rdom, tmp3[rdom, xindex])
        tmp5 = hl.Func('tmp5')
        tmp5[rindex, xindex] = tmp3[rindex, xindex] / tmp4[xindex]
        out_ptr3_i0 = hl.Var('out_ptr3_i0')
        out_ptr3_i1 = hl.Var('out_ptr3_i1')
        out_ptr3[out_ptr3_i0, out_ptr3_i1] = hl.cast(out_ptr3.type(), tmp5[out_ptr3_i0, out_ptr3_i1])

        assert g.using_autoscheduler()
        in_ptr0.set_estimates([hl.Range(0, 512)])
        out_ptr3.set_estimates([hl.Range(0, 32), hl.Range(0, 16)])
```

After
```py
@hl.generator(name="kernel")
class Kernel:
    in_ptr0 = hl.InputBuffer(hl.Float(32), 2)
    out_ptr3 = hl.OutputBuffer(hl.Float(32), 2)

    def generate(g):
        in_ptr0 = g.in_ptr0
        out_ptr3 = g.out_ptr3
        h0 = hl.Var('h0')
        h1 = hl.Var('h1')
        rdom = hl.RDom([hl.Range(0, 32)])
        hr1 = rdom[0]
        tmp0 = hl.Func('tmp0')
        tmp0[h0, h1] = in_ptr0[h0, h1,]
        tmp1 = hl.Func('tmp1')
        tmp1[h1] = hl.maximum(rdom, tmp0[hr1, h1])
        tmp2 = hl.Func('tmp2')
        tmp2[h0, h1] = tmp0[h0, h1] - tmp1[h1]
        tmp3 = hl.Func('tmp3')
        tmp3[h0, h1] = hl.fast_exp(hl.cast(hl.Float(32), tmp2[h0, h1])) if tmp2.type().bits() <= 32 else hl.exp(tmp2[h0, h1])
        tmp4 = hl.Func('tmp4')
        tmp4[h1] = hl.sum(rdom, tmp3[hr1, h1])
        tmp5 = hl.Func('tmp5')
        tmp5[h0, h1] = tmp3[h0, h1] / tmp4[h1]
        out_ptr3[h0, h1,] = hl.cast(hl.Float(32), tmp5[h0, h1])

        assert g.using_autoscheduler()
        in_ptr0.dim(0).set_min(0)
        in_ptr0.dim(0).set_stride(1)
        in_ptr0.dim(0).set_extent(32)
        in_ptr0.dim(1).set_min(0)
        in_ptr0.dim(1).set_stride(32)
        in_ptr0.dim(1).set_extent(16)
        in_ptr0.set_estimates([hl.Range(0, 32), hl.Range(0, 16)])
        out_ptr3.set_estimates([hl.Range(0, 32), hl.Range(0, 16)])
```

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/codegen/halide.py [hunk 1]
```diff
@@ -528,39 +545,65 @@ class HalideCSEVariable(CSEVariable):
...
+
+@dataclasses.dataclass
+class DimensionInfo:
+    expr: Optional[sympy.Expr]
+    size: sympy.Expr
+    stride: sympy.Expr
+
+    def __init__(self, expr, size, stride):
+        super().__init__()
+        if V.graph.sizevars.statically_known_leq(stride, 0):
```
Comment (shunting314) [inline torch/_inductor/codegen/halide.py:585]: hmm, when do we get a negative stride?

File: torch/_inductor/codegen/halide.py [hunk 2]
```diff
@@ -591,42 +634,350 @@ def __init__(
         self.indexing_code_dom = IndentedBuffer()
         self.needs_dom_indexing = self.inside_reduction
         self.has_reduction = self.inside_reduction
-        self.store_buffer_dimensions: Dict[str, List[sympy.Expr]] = {}
+        self.buffer_dimensions: Dict[str, List[DimensionInfo]] = {}
+        self.buffer_offsets: Dict[str, sympy.Expr] = {}
+        self.halide_vars: Dict[sympy.Symbol, sympy.Expr] = {}
+        self.index_replacements: Dict[sympy.Expr, sympy.Expr] = {}
+        self.reduction_renames: Dict[sympy.Symbol, sympy.Symbol] = {}
+        self.dom_renames: Dict[str, Dict[sympy.Symbol, sympy.Symbol]] = {}
```
Comment (shunting314) [inline torch/_inductor/codegen/halide.py:642]: It's hard to understand at a first glance what the keys for the outer and inner dict means. Add some comment?

File: torch/_inductor/codegen/halide.py [hunk 3]
```diff
@@ -591,42 +634,350 @@ def __init__(
...
+        we base indexing on a larger number of vars whose product combines to those.
+
+        This function populates self.halide_vars, self.index_replacements, and self.reduction_renames
+        """
+        assert not (
+            self.index_replacements or self.halide_vars or self.reduction_renames
+        )
+        eq = V.graph.sizevars.statically_known_equals
+        lt = V.graph.sizevars.statically_known_lt
+        size_hint = functools.partial(V.graph.sizevars.size_hint, fallback=inf)
```
Comment (shunting314) [inline torch/_inductor/codegen/halide.py:665]: Should we use an integer rather than a float for the fallback value?

**Review by shunting314 — APPROVED** [review]

File: torch/_inductor/codegen/halide.py (hunk 3 above)
Comment (jansel) [reply torch/_inductor/codegen/halide.py:665]: I want it to go last and there is no such thing as a max int in python

File: torch/_inductor/codegen/halide.py (hunk 1 above)
Comment (jansel) [reply torch/_inductor/codegen/halide.py:585]: x[::-1]

File: torch/_inductor/codegen/halide.py [hunk 4]
```diff
@@ -640,175 +1038,94 @@ def used_dims_from_index(self, index: sympy.Expr):
...
-        if self.is_indirect_indexing(index) or self._load_mask:
-            # Halide doesn't have a great way to do masked loads
-            var = f"hl.BoundaryConditions.constant_exterior({var}, 0)"
-        line = f"{var}[{index_str}]"
+        var, dims = self.indexing_to_dimensions(var, index, False)
+        index_str = ", ".join(d.index_str() for d in dims)
+        line = f"{var}[{index_str},]"  # trailing comma workaround for https://github.com/halide/Halide/issues/8299
         dtype = V.graph.get_dtype(name)
         if dtype in (torch.float16, torch.bfloat16):
+            dtype = torch.float32
```
Comment (eellison) [inline torch/_inductor/codegen/halide.py:1072]: nit: factor out to `dtype_to_compute_dtype` similar to triton codegen ?

File: torch/_inductor/codegen/halide.py [hunk 5]
```diff
@@ -591,41 +663,367 @@ def __init__(
...
+                    FloorDiv(
+                        sympy.Wild("base"),
+                        sympy.Wild("divisor"),
+                    ),
+                    visit_floor_div,
+                )
+            all_used_symbols.update(super().prepare_indexing(index).free_symbols)
+
+        had_fallback = False
+        for tree in reversed(self.range_trees):
```
Comment (eellison) [inline torch/_inductor/codegen/halide.py:756]: nit: maybe factor out to helper function

**Review by eellison — APPROVED** [review]

File: torch/_inductor/codegen/halide.py [hunk 6]
```diff
@@ -1090,15 +1405,25 @@ def update_index(m):
...
+                                f"{arg.name}.dim({i}).set_stride({int(dim.stride)})"
+                            )
+                        except TypeError:
+                            pass  # not integer
+                        try:
+                            code.writeline(
+                                f"{arg.name}.dim({i}).set_extent({int(dim.size)})"
+                            )
+                        except TypeError:
+                            pass  # not integer
```
[the remark this answers is not in the document]
Comment (jansel) [reply torch/_inductor/codegen/halide.py:1428]: It might be a regular int (not sympy)

File: torch/_inductor/codegen/halide.py (hunk 3 above)
Comment (jansel) [reply torch/_inductor/codegen/halide.py:702]: Yeah I suppose, seems like a style preference.

**Pull request closed without merging**